orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -90,24 +89,39 @@ IMPLEMENT_GRADIENT_BUILDER(GetLogGradient) {
 }
 
 IMPLEMENT_GRADIENT_BUILDER(GetTanhGradient) {
-  return std::vector<NodeDef>{
-      NodeDef("TanhGrad",
-              {O(0), GO(0)},
-              {GI(0)})};
+  ArgDef Y = O(0);
+  std::vector<NodeDef> result;
+  NodeDef one_constant_node = OneConstantNode(OElemType(0));
+  ArgDef one_arg = one_constant_node.output_args[0];
+  result.push_back(one_constant_node);
+  result.push_back(NodeDef("Mul", {Y, Y}, {IA("Squared_Y")}));
+  result.push_back(NodeDef("Sub", {one_arg, IA("Squared_Y")}, {IA("Sub_Squared_Y")}));
+  result.push_back(NodeDef("Mul", {GO(0), IA("Sub_Squared_Y")}, {GI(0)}));
+  return result;
```

Comment:
I had a discussion with Sherlock before on this. Since the implementation needs OneConstantNode, and when moving gradient builder after mixed precision, we need to pass in the element type as parameter. If we still use function body to do this, we need to add a new attribute to the TanhGrad schema for this purpose. Moving the implementation here seems more clean and clear. If in the future we need a kernel install of such sub-graph, we can change back.

Reply:
The original design was decided to associate the function body of grad of an op with a subgraph which is composed by a few primitive ops. If there is no native grad for that op then the function body will be inlined.  The change of using function body was initially done by @souptc. So, I would leave it to him to comment.